interface/ceed-elemrestriction.c

```diff
@@ -178,39 +178,33 @@ const CeedElemRestriction CEED_ELEMRESTRICTION_NONE =
   @brief Create a CeedElemRestriction
 
   @param ceed       A Ceed object where the CeedElemRestriction will be created
-  @param imode      Ordering of the ncomp components, i.e. it specifies
-                      the ordering of the components of the L-vector used
-                      by this CeedElemRestriction. CEED_NONINTERLACED indicates
-                      the component is the outermost index and CEED_INTERLACED
-                      indicates the component is the innermost index in
-                      ordering of the L-vector.
-  @param nelem      Number of elements described in the @a indices array
+  @param nelem      Number of elements described in the @a offsets array
   @param elemsize   Size (number of "nodes") per element
-  @param nnodes     The number of nodes in the L-vector. The input CeedVector
-                      to which the restriction will be applied is of size
-                      @a nnodes * @a ncomp. This size may include data
-                      used by other CeedElemRestriction objects describing
-                      different types of elements.
   @param ncomp      Number of field components per interpolation node
                       (1 for scalar fields)
-  @param mtype      Memory type of the @a indices array, see CeedMemType
-  @param cmode      Copy mode for the @a indices array, see CeedCopyMode
-  @param indices    Array of shape [@a nelem, @a elemsize]. Row i holds the
-                      ordered list of the indices (into the input CeedVector)
+  @param compstride Stride between components for the same L-vector "node".
+                      Data for node i, component k can be found in the L-vector
+                      at index [offsets[i] + k*compstride].
+  @param lsize      The size of the L-vector. This vector may be larger than
+                      the elements and fields given by this restriction.
+  @param mtype      Memory type of the @a offsets array, see CeedMemType
+  @param cmode      Copy mode for the @a offsets array, see CeedCopyMode
+  @param offsets    Array of shape [@a nelem, @a elemsize]. Row i holds the
+                      ordered list of the offsets (into the input CeedVector)
                       for the unknowns corresponding to element i, where
-                      0 <= i < @a nelem. All indices must be in the range
-                      [0, @a nnodes - 1].
+                      0 <= i < @a nelem. All offsets must be in the range
+                      [0, @a lsize - 1].
```

Comment:
Do you have code that checks whether this criteria is satisfied?

Reply:
Do we? No. Can we? Yes.

It would be easy to check. It seems like this would be nice to put in the debugging mode we've been planning? It's only checking `nelem*elemsize` values, so I can put it in this PR if its desired and turn if off by default for the non debugging mode.

Reply:
Just do it all the time for reference backends?

Reply:
Right now the ref CPU backend restriction is used for all optimized CPU backends.  I guess I could turn on this check only when the backend resource name contains `ref`?

Reply:
Yeah, or a private flag that you set to enable that sort of checking behavior (so you aren't checking resource names during execution).

Reply:
What about the added check? It's only a strcmp on creation of a restriction and offsets verification if we have a ref or memcheck backend.